### ISSUE TITLE
Ability to manully set the IP address

### ISSUE THREAD

**Issue by Benigans** [kovapatrik/homebridge-midea-platform#35]
Hi !
Since the last update, I can't set an IP address in the plugin config.
My dehumidifier is in an IOT VLAN and the discovery doesn't work (homebridge on another VLAN).

With previous plugin version, I could set a fixed IP address and that worked.
With the latest, I can only set ID, Token and Key. And it doesn't work for me.

If I use the midea-beautiful-air-cli command to get status of my dehumidifier (by putting the IP address as a parameter), I can get the status.

Could you add the possibility to force the IP address of the dehumidifier ?

Thx !

**Comment by kovapatrik** [comment]
That's an edge case I didn't think about before. Is it possible to add a network interface where your Homebridge instance is running? That would be the easiest solution in my opinion.

If not, I could add a new button in the GUI, where you can explicitly specifiy on what IP you want to search for a device.

**Comment by dkerr64** [comment]
@Benigans Thanks for testing and pointing this out.  I have a IOT VLAN myself, actually several VLANs for different types of devices (cameras on one, other home automation on another, etc.) and found that for Homebridge to work best I needed to create network interfaces for each on my Homebridge server (virtual machine).  This ensures that broadcasts get to all my subnets.

However, I can understand if this is not ideal or may not even be possible in your network design.  So net-net, I think we need to add back in ability to specify a device IP address in the config, and a way to discover the token/key from a specific IP address in the setting GUI window.

Thanks

**Comment by kovapatrik** [comment]
It's kind of an 'expert usage' if someone has their devices on a subnetwork in my opinion, but I might be wrong here. But if you agree then I would expect such a user to create interfaces. In addition we can add some instructions on how to do this.

**Comment by kovapatrik** [comment]
But yeah, we can also add the IP field with very specific instructions for when to use it. As for the GUI part, I would create a seperate button for it.

**Comment by dkerr64** [comment]
I'm assuming here that traffic between @Benigans server to his IOT VLAN has to be routed through a network router.  That was in fact how my network was originally setup as it is the easy way to do it without over thinking things.  Unfortunately routers do not forward broadcasts, they stay only on the local subnet.

I had to rethink my network design and add interfaces for specific VLAN subnets to my Homebridhge server.   But this is not as easy as it sounds.  And if you are on older virtualization platforms may not even be possible (LXC.. linux containers... only recently (last couple years) added ability to pass through VLANs to containers).

In other words, I think we need to handle ourselves.  Adding specific IP back to config.json and handling in the plugin initialization is easy, we had that before.

For GUI discovery, I would suggest an entry field to left of the discover all devices button.

`IP: (optional text field) [Discover all devices button]`

If entry field is empty, broadcast as normal. If a IP address is provided, unicast to that one address.  It does mean that if you have multiple devices you will have to change that field, discover again, etc.  But I see no avoiding that.

**Comment by kovapatrik** [comment]
Another solution would be the ability to add subnets into the plugin. This way only minimal modification is needed as only the `Discover` class would change.

**Comment by Benigans** [comment]
Thx guys !
I'll try to add a network interface to my homebridge.

**Comment by dkerr64** [comment]
Thinking about this overnight...

- We should add an ip field to advanced_options in the UI schema  At plugin startup this is handled the way we used to do it.
- In the UI, with the field in the schema, a user can add devices manually and just add an IP address.  Then on pressing Discover All Devices button we send unicast to IP addresses configured.  And we handle add/update just as we do today with exception that on "add device" (available when no device ID is in the config) we need to check if the IP address is listed.
- And with storing IP address in the config, on "update device" we need to update the IP address as well as token/key.

**Comment by kovapatrik** [comment]
If there is no other solution then I mostly agree with this. I'm not familiar with subtnetworks to be honest so I have a question: is it possble to send broadcast message to a subnetwork, even if a network intreface is not existing for it? Because if it is, then I would simply just add explicit IPs into the discovery "interfaces".

**Comment by dkerr64** [comment]
> If there is no other solution then I mostly agree with this. I'm not familiar with subtnetworks to be honest so I have a question: is it possble to send broadcast message to a subnetwork, even if a network intreface is not existing for it? Because if it is, then I would simply just add explicit IPs into the discovery "interfaces".

Yes you can broadcast to a subnet, in fact that is what we do.  You have two options... 255.255.255.255 or say 192.168.1.255 both of those are broadcast addresses.  However in both cases, at the source device, the broadcast is only sent out on one interface.  If you use 255.255.255.255 it is only sent on the "first" interface, as determined by the operating system.

If you have multiple interfaces then you can't be sure for certain which one the OS will select.  So instead you find the subnets of each interface and broadcast to X.Y.Z.255 (or whatever the broadcast address for the subnet is based on the net mask.  It could be 10.255.255.255 for example) for each interface.

The problem with broadcast addresses is that they are not routed.  They stay on your local subnet.  So even if someone specifically addresses 192.168.1.255, it will not get to anything if you are on 192.168.100.1/24

**Comment by dkerr64** [comment]
@kovapatrik have you started a branch to work on this?  If not I will make a start on it.

**Comment by kovapatrik** [comment]
> @kovapatrik have you started a branch to work on this? If not I will make a start on it.

No, I did not.

**Comment by dkerr64** [comment]
I have made a start [here](https://github.com/dkerr64/homebridge-midea-platform/commit/eee11ffca50a5354bd65063ef5d0d82f8f7e7b2e).  I have IP address added to schema and plugin initialization will use it.  So can manually add IP addresses.   Next step is to discover the token/key credentials in the user interface.

**Comment by dkerr64** [comment]
PR #38 opened, please review.

**Comment by dkerr64** [comment]
@Benigans  please try the just published release.  You can put the IP addresses of your devices into the "discover devices" user interface and it should find them.  A space- or comma-separated list of your known IP addresses.

Thanks.

**Comment by Benigans** [comment]
> @Benigans please try the just published release. You can put the IP addresses of your devices into the "discover devices" user interface and it should find them. A space- or comma-separated list of your known IP addresses.
> 
> Thanks.

It’s working perfectly 👌🏻
Thx for the update !
